### PR TITLE
Adds Rage Fist

### LINKED ISSUE #2507: MOVE_RAGE_FIST Effect

### PR DESCRIPTION
**Pull request by AlexOn1ine**
Closes #2507

### CODE REVIEW

**Comment by AlexOn1ine** [comment]
Ready for review.

I initial followed [Bulbagarden](https://bulbapedia.bulbagarden.net/wiki/Rage_Fist_(move)) but switched to [Showdown](https://dex.pokemonshowdown.com/moves/ragefist) because I think their implementation is correct and was tested by [DaWoblefet](https://www.youtube.com/watch?v=Qu3NkYloI_Y)

**Comment by Bassoonian** [comment]
You've got a lot of failing tests here, so probably best to fix those first :)

**Comment by AlexOn1ine** [comment]
> You've got a lot of failing tests here, so probably best to fix those first :)

Messed up merge 😅

Also I forgot to mention. I added callnative `BS_TryCopycat`. I thought I'll need it for the Rage Fist effect. Wasn't the case but since I already changed it, there was no point in reverting it back to a various.

File: include/battle.h [hunk 1]
```diff
@@ -731,6 +731,7 @@ struct BattleStruct
     u32 aiDelayFrames; // Number of frames it took to choose an action.
     bool8 transformZeroToHero[PARTY_SIZE][NUM_BATTLE_SIDES];
     u8 pledgeMove:1;
+    u32 timesGotHit[PARTY_SIZE][NUM_BATTLE_SIDES];
```
Comment (Skeli789) [inline include/battle.h:734]: Can be `u8` since the damage is capped

File: src/battle_script_commands.c [hunk 2]
```diff
@@ -2243,6 +2243,11 @@ static void Cmd_healthbarupdate(void)
             if (GetBattlerSide(battler) == B_SIDE_PLAYER && gBattleMoveDamage > 0)
                 gBattleResults.playerMonWasDamaged = TRUE;
         }
+        if (!(gHitMarker & HITMARKER_PASSIVE_DAMAGE) && !DoesSubstituteBlockMove(gBattlerAttacker, battler, gCurrentMove)
+            && gBattleStruct->timesGotHit[gBattlerPartyIndexes[battler]][GetBattlerSide(battler)] < 6)
+        {
+            gBattleStruct->timesGotHit[gBattlerPartyIndexes[battler]][GetBattlerSide(battler)]++;
```
Comment (Skeli789) [inline src/battle_script_commands.c:2249]: I recommend not putting this here and instead adding it to the same place the logic for `MOVE_RAGE` is handled, considering that Rage Fist is just a buffed version of Rage.

File: src/battle_util.c [hunk 3]
```diff
@@ -8862,6 +8862,9 @@ static inline u32 CalcMoveBasePower(u32 move, u32 battlerAtk, u32 battlerDef, u3
     case EFFECT_MAX_MOVE:
         basePower = GetMaxMovePower(gBattleMons[battlerAtk].moves[gBattleStruct->chosenMovePositions[battlerAtk]]);
         break;
+    case EFFECT_RAGE_FIST:
+        basePower += 50 * gBattleStruct->timesGotHit[gBattlerPartyIndexes[battlerAtk]][GetBattlerSide(battlerAtk)];
```
Comment (Skeli789) [inline src/battle_util.c:8866]: The damage should be capped at 350

**Review by Skeli789 — CHANGES_REQUESTED** [review]

File: src/battle_util.c (hunk 3 above)
Comment (kleeenexfeu) [reply src/battle_util.c:8866]: The counter is limited at 6, so it is accounted for

Comment (Skeli789) [reply src/battle_util.c:8866]: The damage and not the counter should be limited. If the field ever needs to be reused in the future and whatever is using it has a different cap, the field value can always be relied on.

Comment (AlexOn1ine) [reply src/battle_util.c:8866]: Yes, this makes sense, will adjust.

**Comment by AlexOn1ine** [comment]
Applied review changes, ready for re-review!

File: src/battle_script_commands.c [hunk 4]
```diff
@@ -5537,6 +5537,16 @@ static void Cmd_moveend(void)
             }
             gBattleScripting.moveendState++;
             break;
+        case MOVEEND_NUM_HITS:
+            if (gBattlerAttacker != gBattlerTarget
+                && GetBattlerSide(gBattlerAttacker) != GetBattlerSide(gBattlerTarget)
+                && !(gMoveResultFlags & MOVE_RESULT_NO_EFFECT) && gBattleMoves[gCurrentMove].power != 0
+                && !DoesSubstituteBlockMove(gBattlerAttacker, gBattlerTarget, gCurrentMove))
```
Comment (schweits) [inline src/battle_script_commands.c:5544]: Missing `TARGET_TURN_DAMAGED` and `gBattleMons[gBattlerTarget].hp != 0`,

File: src/battle_script_commands.c [hunk 5]
```diff
@@ -5537,6 +5537,16 @@ static void Cmd_moveend(void)
             }
             gBattleScripting.moveendState++;
             break;
+        case MOVEEND_NUM_HITS:
+            if (gBattlerAttacker != gBattlerTarget
```
Comment (Skeli789) [inline src/battle_script_commands.c:5541]: Missing `TARGET_TURN_DAMAGED` and `gBattleMons[gBattlerTarget].hp != 0`

Comment (AlexOn1ine) [reply src/battle_script_commands.c:5541]: > `TARGET_TURN_DAMAGED`

As far as I understand it you don't need to take damage as long as you get hit by a damage dealing move (e.g. False Swipes).

> `gBattleMons[gBattlerTarget].hp != 0`

What is the reason for this check?

Comment (Skeli789) [reply src/battle_script_commands.c:5541]: > As far as I understand it you don't need to take damage as long as you get hit by a damage dealing move (e.g. False Swipes).
`TARGET_TURN_DAMAGED` should still evaluate to true for False Swipe. If it doesn't that's a bug. You need it because otherwise this branch will be triggered by Pain Split (by virtue of the power being set to 1) even though it doesn't do direct damage.

> What is the reason for this check?
Looked it up, and you actually shouldn't add it in the end. The counter's not reset upon fainting like with othercounters.

Comment (AlexOn1ine) [reply src/battle_script_commands.c:5541]: > `TARGET_TURN_DAMAGED` should still evaluate to true for False Swipe. If it doesn't that's a bug. You need it because otherwise this branch will be triggered by Pain Split (by virtue of the power being set to 1) even though it doesn't do direct damage.

Not sure what to do about `TARGET_TURN_DAMAGED`. It checks for non zero damage so I would need a workaround for False Swipe, Endure and Focus Band since all of those can result in 0 damage, right?

Comment (Skeli789) [reply src/battle_script_commands.c:5541]: You should double check if it's actually not evaluating to true for False Swipe. Because if it is, certain branches of code like:
```
        case ABILITY_STATIC:
            if (!(gMoveResultFlags & MOVE_RESULT_NO_EFFECT)
             && gBattleMons[gBattlerAttacker].hp != 0
             && !gProtectStructs[gBattlerAttacker].confusionSelfDmg
             && TARGET_TURN_DAMAGED
             && CanBeParalyzed(gBattlerAttacker)
             && IsMoveMakingContact(move, gBattlerAttacker)
             && RandomWeighted(RNG_STATIC, 2, 1))
```

Will not be triggered in those cases you mentioned even though they very much should be.

Comment (Skeli789) [reply src/battle_script_commands.c:5541]: And if it is indeed not, look at how setting those values is handled in the CFRU. They're always set to at least `1` if the target is damaged.

Comment (AlexOn1ine) [reply src/battle_script_commands.c:5541]: Yep, Static and False Swipe interaction is also bugged. 

As far as I can tell `gHpDealt` is set to `gBattleMoveDamage` and I don't see where `physicalDmg` could be set to 1 if False Swipes deals 0 damage. 
```
if (gBattleMons[battler].hp > gBattleMoveDamage)
{
    gBattleMons[battler].hp -= gBattleMoveDamage;
    gHpDealt = gBattleMoveDamage;
}
``` 
I looked into CFRU and unless I'm misunderstand it the only time `hpDealt` can be set to 1 is if Raid Shields are up
```
if (HasRaidShields(gActiveBattler)) //Raid Shields can be used outside of Raid Battles now
    hpDealt = MathMax(1, hpDealt); //Because damage can get heavily reduced to 0
```
Unless I'm looking at the wrong place but I don't know why `gBattleMoveDamage` would be set to 1 in this case when `Cmd_datahpupdate` is called.

File: src/battle_script_commands.c [hunk 6]
```diff
@@ -5537,6 +5538,16 @@ static void Cmd_moveend(void)
...
+                && GetBattlerSide(gBattlerAttacker) != GetBattlerSide(gBattlerTarget)
```
Comment (Skeli789) [inline src/battle_script_commands.c:5543]: Remove this and the line above. Confusion damage counts. The three checks after this are enough (although I would move the `gBattleMoves[gCurrentMove].split != SPLIT_STATUS` to it's own line)

File: include/battle.h [hunk 7]
```diff
@@ -732,6 +732,8 @@ struct BattleStruct
     bool8 transformZeroToHero[PARTY_SIZE][NUM_BATTLE_SIDES];
     u8 pledgeMove:1;
     bool8 isSkyBattle:1;
+    u8 timesGotHit[PARTY_SIZE][NUM_BATTLE_SIDES];
```
Comment (Skeli789) [inline include/battle.h:735]: Personally I would put the battle side first and make the party size the subarray, but I see you're just following `transformZeroToHero` which was also done this way

**Review by Skeli789 — CHANGES_REQUESTED** [review]

File: src/battle_script_commands.c (hunk 6 above)
Comment (AlexOn1ine) [reply src/battle_script_commands.c:5543]: > Confusion damage counts
Does it? I know Bulbagarden says it does but both Showdown and research from DaWoblefet say it doesn't.

File: include/battle.h (hunk 7 above)
Comment (AlexOn1ine) [reply include/battle.h:735]: Out of curiosity. Why do you think it's better? I wouldn't mind changing it though this is what all fields look like in the expansion if they need party and side.

Comment (AlexOn1ine) [reply include/battle.h]: Yeah I agree and also changed it since it looks like there is no consistency right now anyway and `transformZeroToHero` field is changed in #3614 so it should be fine.

File: src/battle_script_commands.c (hunk 6 above)
Comment (AlexOn1ine) [reply src/battle_script_commands.c]: All done. 

For Substitute I already had a test specifically to test the edge case.

**Review by Bassoonian — APPROVED** [review]
